python-spec/src/somacore/query/fast_csr.py

```diff
@@ -0,0 +1,299 @@
+    data: npt.NDArray[np.number],
+    indices: npt.NDArray[np.signedinteger],
+    indptr: npt.NDArray[np.signedinteger],
+    row_rng_mask: int,
+    row_rng_val: int,
+):
+    for n in range(len(data_chunk)):
+        row = row_ind_chunk[n]
+        if (row & row_rng_mask) != row_rng_val:
+            continue
+        ptr = indptr[row]
+        indices[ptr] = col_ind_chunk[n]
+        data[ptr] = data_chunk[n]
+        indptr[row] += 1
+
+
+@numba.jit(nopython=True, nogil=True)  # type: ignore
+def _copy_chunklist_range(
+    chunk_list: numba.typed.List,
+    data: npt.NDArray[np.number],
+    indices: npt.NDArray[np.signedinteger],
+    indptr: npt.NDArray[np.signedinteger],
+    row_rng_mask_bits: int,
+    job: int,
+):
+    row_rng_mask = (2**64 - 1) >> row_rng_mask_bits << row_rng_mask_bits
+    row_rng_val = job << row_rng_mask_bits
+    for data_chunk, row_ind_chunk, col_ind_chunk in chunk_list:
+        _copy_chunk_range(
+            data_chunk,
+            row_ind_chunk,
+            col_ind_chunk,
+            data,
+            indices,
+            indptr,
+            row_rng_mask,
+            row_rng_val,
+        )
+
+
+@numba.jit(nopython=True, nogil=True)  # type: ignore
+def _finalize_indptr(indptr: npt.NDArray[np.signedinteger]):
+    prev = 0
+    for r in range(len(indptr)):
+        t = indptr[r]
+        indptr[r] = prev
+        prev = t
+
+
+def _select_dtype(
+    maxval: int,
+) -> Type[np.signedinteger]:
+    """
+    Ascertain the "best" dtype for a zero-based index. Given our
+    goal of minimizing memory use, "best" is currently defined as
+    smallest.
+    """
+    if maxval > np.iinfo(np.int32).max:
+        return np.int64
+    else:
+        return np.int32
+
+
+def _reindex_and_cast(
+    index: pd.Index, ids: npt.NDArray[np.int64], target_dtype: npt.DTypeLike
+) -> npt.NDArray[np.int64]:
+    return cast(
+        npt.NDArray[np.int64], index.get_indexer(ids).astype(target_dtype, copy=False)
+    )
+
+
+def _read_csr(
+    matrix: scd.SparseNDArray, obs_joinids: pa.Array, var_joinids: pa.Array
+) -> Tuple[
+    npt.NDArray[np.number],  # data
+    npt.NDArray[np.integer],  # indptr
+    npt.NDArray[np.integer],  # indices
+    Tuple[int, int],  # shape
+]:
+    if not isinstance(matrix, scd.SparseNDArray) or matrix.ndim != 2:
+        raise TypeError("read_scipy_csr can only read from a 2D SparseNDArray")
+
+    max_workers = (os.cpu_count() or 4) + 2
+    with futures.ThreadPoolExecutor(max_workers=max_workers) as pool:
+        acc = _CSRAccumulator(
+            obs_joinids=obs_joinids, var_joinids=var_joinids, pool=pool
+        )
+        for tbl in eager_iter._EagerIterator(
+            matrix.read((obs_joinids, var_joinids)).tables(),
+            pool=pool,
+        ):
+            acc.append(tbl["soma_data"], tbl["soma_dim_0"], tbl["soma_dim_1"])
+
+        data, indptr, indices, shape = acc.finalize()
+
+    return data, indptr, indices, shape
+
+
+def read_scipy_csr(
```

Comment:
It would do a lot for readability to put the main entrypoint methods (`read_scipy_csr` and `read_arrow_csr`) at the top of the file -- which is where people generally begin reading.